python/instrumentation/openinference-instrumentation-llama-index/tests/openinference/instrumentation/llama_index/test_callback.py

```diff
@@ -0,0 +1,354 @@
+    openai_span = spans_by_name.pop(next(iter(spans_by_name.keys())))
+    if is_stream:
+        # FIXME: for streaming, the openai call happens after the trace has ended, so it's orphaned.
+        # https://github.com/run-llama/llama_index/blob/fcfab6486bc6a0eec31a983dd3056ef9cbe8ceb2/llama_index/llms/base.py#L62  # noqa E501
+        assert openai_span.parent is None
+    else:
+        assert openai_span.parent is not None
+        if status_code == 200:
+            # FIXME: LlamaIndex doesn't currently capture the LLM span when status_code == 400
+            # because `on_event_end` never gets called.
+            assert llm_span is not None
+            assert openai_span.parent.span_id == llm_span.context.span_id
+            assert openai_span.context.trace_id == llm_span.context.trace_id
+
+    assert spans_by_name == {}
+
+
+@pytest.fixture
+def nodes() -> List[TextNode]:
+    return [
+        Document(text=randstr(), metadata={"category": randstr()}),
+        TextNode(text=randstr()),
+    ]
+
+
+@pytest.fixture
+def chat_completion_mock_stream() -> Tuple[List[bytes], List[Dict[str, Any]]]:
+    return (
+        [
+            b'data: {"choices": [{"delta": {"role": "assistant"}, "index": 0}]}\n\n',
+            b'data: {"choices": [{"delta": {"content": "A"}, "index": 0}]}\n\n',
+            b'data: {"choices": [{"delta": {"content": "B"}, "index": 0}]}\n\n',
+            b'data: {"choices": [{"delta": {"content": "C"}, "index": 0}]}\n\n',
+            b"data: [DONE]\n",
+        ],
+        [{"role": "assistant", "content": "ABC"}],
+    )
+
+
+@pytest.fixture(scope="module")
+def in_memory_span_exporter() -> InMemorySpanExporter:
+    return InMemorySpanExporter()
+
+
+@pytest.fixture(scope="module")
+def tracer_provider(in_memory_span_exporter: InMemorySpanExporter) -> trace_api.TracerProvider:
+    resource = Resource(attributes={})
+    tracer_provider = trace_sdk.TracerProvider(resource=resource)
+    span_processor = SimpleSpanProcessor(span_exporter=in_memory_span_exporter)
+    tracer_provider.add_span_processor(span_processor=span_processor)
+    return tracer_provider
+
+
+@pytest.fixture(autouse=True)
+def instrument(
+    tracer_provider: trace_api.TracerProvider,
+    in_memory_span_exporter: InMemorySpanExporter,
+) -> Generator[None, None, None]:
+    LlamaIndexInstrumentor().instrument(tracer_provider=tracer_provider)
+    OpenAIInstrumentor().instrument(tracer_provider=tracer_provider)
+    yield
+    OpenAIInstrumentor().uninstrument()
+    LlamaIndexInstrumentor().uninstrument()
+    in_memory_span_exporter.clear()
+
+
+@pytest.fixture(autouse=True)
+def openai_api_key(monkeypatch: pytest.MonkeyPatch) -> None:
+    monkeypatch.setenv("OPENAI_API_KEY", "sk-")
+
+
+@pytest.fixture(scope="module")
+def seed() -> Iterator[int]:
+    return count()
+
+
+@pytest.fixture(autouse=True)
+def set_seed(seed: Iterator[int]) -> Iterator[None]:
+    random.seed(next(seed))
+    yield
```

Comment:
This way of seeding leads to the tests having a different random seed if a different subset of the tests are run. For reproducibility, we might just leave the random seed as a fixed number.

Reply:
That's right. But it will still be consistent when the subset is fixed. I think reproducibility here is defined in terms of the fixed set of tests that needs to be run. Because if an error occurs you're likely to just run the same test over and over again, and in that scenario, the seed will work as intended.

Reply:
discussed offline